### ISSUE TITLE
gRPC requests fail if made via nginx k8s ingress controller

### ISSUE THREAD

**Issue by maloo** [dotnet/aspnetcore#46079]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

- I have followed the standard instructions on how to use gRPC behind nginx from here: https://kubernetes.github.io/ingress-nginx/examples/grpc/#step-3-create-the-kubernetes-ingress-resource-for-the-grpc-app . The only real change from any normal ingress is that backend-protocol is specified as GRPC.
- My dotnet 7 gRPC app is deplyed and both server and client working via kubectl port-forward to the service port of the pod.
- Server app has set kestrel to use only http2 on port 80. Allowing http1.x fails in unsecure/http mode which nginx use (backend-protocol grpc).
- http/2 is working via nginx using "curl --http2-prior-knowledge ..." to a hello world GET / endpoint

So everything looks ok, except when I try a dotnet client to call the grpc service I get the following error:
```
Unhandled exception: Grpc.Core.RpcException: Status(StatusCode="Unavailable", Detail="Error starting gRPC call. HttpRequestException: An error occurred while sending the request. IOException: The request was aborted. IOException: An HTTP/2 connection could not be established because the server did not complete the HTTP/2 handshake. HttpProtocolException: The HTTP/2 server sent invalid data on the connection. HTTP/2 error code 'PROTOCOL_ERROR' (0x1).", DebugException="System.Net.Http.HttpRequestException: An error occurred while sending the request.
 ---> System.IO.IOException: The request was aborted.
 ---> System.IO.IOException: An HTTP/2 connection could not be established because the server did not complete the HTTP/2 handshake.
 ---> System.Net.Http.HttpProtocolException: The HTTP/2 server sent invalid data on the connection. HTTP/2 error code 'PROTOCOL_ERROR' (0x1).
   at System.Net.Http.Http2Connection.ThrowProtocolError(Http2ProtocolErrorCode errorCode, String message)
   at System.Net.Http.Http2Connection.ReadFrameAsync(Boolean initialFrame)
   at System.Net.Http.Http2Connection.ProcessIncomingFramesAsync()
   --- End of inner exception stack trace ---
   at System.Net.Http.Http2Connection.ProcessIncomingFramesAsync()
   --- End of inner exception stack trace ---
   at System.Net.Http.Http2Connection.ThrowRequestAborted(Exception innerException)
   at System.Net.Http.Http2Connection.Http2Stream.CheckResponseBodyState()
   at System.Net.Http.Http2Connection.Http2Stream.TryEnsureHeaders()
   at System.Net.Http.Http2Connection.Http2Stream.ReadResponseHeadersAsync(CancellationToken cancellationToken)
   at System.Net.Http.Http2Connection.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.Http2Connection.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithVersionDetectionAndRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at Grpc.Net.Client.Internal.GrpcCall`2.RunCall(HttpRequestMessage request, Nullable`1 timeout)")
   at UnitListCommand.GenerateSplits(Uri server) in C:\Data\Projects\GitHub\NT-SplitServer\ntuclient\UnitListCommand.cs:line 23
   at System.CommandLine.Invocation.AnonymousCommandHandler.InvokeAsync(InvocationContext context)
   at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass17_0.<<UseParseErrorReporting>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass12_0.<<UseHelp>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass22_0.<<UseVersionOption>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass19_0.<<UseTypoCorrections>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseSuggestDirective>b__18_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass16_0.<<UseParseDirective>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<RegisterWithDotnetSuggest>b__5_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass8_0.<<UseExceptionHandler>b__0>d.MoveNext()
```

Are there any samples of how to deploy a grpc service in kubernetes written in asp.net core that is accessable via an ingress controller, in this case nginx?

### Expected Behavior

No exception of protocol error at grpc request.

### Exceptions (if any)

See bug description above

### .NET Version

7.0.102

### Anything else?

    <PackageReference Include="Google.Protobuf" Version="3.17.3" />
    <PackageReference Include="Grpc.Net.Client" Version="2.39.0" />
    <PackageReference Include="Grpc.Tools" Version="2.39.1">

**Comment by Tratcher** [comment]
What do the server logs from the backend say?

**Comment by maloo** [comment]
TL;DR this can be closed.

I got something like `"PRI * HTTP/2.0" 400 [] []` (from memory). But it turned out that the issue was that I forgot to change client from http to https after testing (and fixing) k8s backend service via port forward (which is http/2, no ssl).

I hope you can create better exceptions and messages when dotnet grpc client connect using http to a server allowing both http/1.x and http/2. It seems to be similar in nginx and asp.net core when using http (in that it will fail if server has http/1.x enabled). Even if this error is not only in this error case, at least add a hint about trying https instead. Message about handshake and protocol errors doesn't give much clues.
